### PR TITLE
[WIP] read_csv - opening file as utf-8-sig

### LINKED ISSUE #544: read_csv - Key 'Name' was not found in the CSV header fields
##### SUMMARY
The `read_csv` module fails to identify a field, yet displaces the field in the list of available fields.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
read_csv

##### ANSIBLE VERSION

```
ansible 2.9.10
  config file = /home/anton/git/ansible-deploy-vmware-vm/ansible.cfg
  configured module search path = ['/home/anton/git/ansible-deploy-vmware-vm/library']
  ansible python module location = /home/anton/.local/lib/python3.6/site-packages/ansible
  executable location = /home/anton/.local/bin/ansible
  python version = 3.6.9 (default, Apr 18 2020, 01:56:04) [GCC 8.4.0]

```

##### CONFIGURATION

```
# config file for ansible -- http://ansible.com/
# ==============================================

# nearly all parameters can be overridden in ansible-playbook
# or with command line flags. ansible will read ANSIBLE_CONFIG,
# ansible.cfg in the current working directory, .ansible.cfg in
# the home directory or /etc/ansible/ansible.cfg, whichever it
# finds first

[defaults]
host_key_checking = False
host_key_check = False
ansible_python_interpreter=/usr/bin/python3
log_path = ./ansible.log
#bin_ansible_callbacks=True
#stdout_callback = debug


# some basic default values...
library        = ./library

# additional paths to search for roles in, colon separated
roles_path    = ./roles

[ssh_connection]
# ssh arguments to use
ssh_args = -o StrictHostKeyChecking=no
timeout=60


```

##### OS / ENVIRONMENT
Ubuntu 20:04

##### STEPS TO REPRODUCE

```
---
- name: Right-size VMs
  gather_facts: false
  hosts: all
  connection: local
  tasks:
    # Read a CSV file and access the first item
  - name: Read users from CSV file and return a list
    read_csv:
        path: "files/vms/6-19-20 Optimization Report - Oversized Virtual Machines Prod2.csv"
        key: Name
    register: users

  - debug:
      msg: 'User {{ users.list.2.Name}}'
        # msg: 'User {{ users.list.2.Name}} has UID {{ users.list.2.ReclaimablevCPUs}} and GID {{ users.list.2.ReclaimableMemory}}'
        # msg: "{{ users }}"



```

##### EXPECTED RESULTS
Expect to be able to read CSV values by col name (field) as based on module documentation.

##### ACTUAL RESULTS
```
fatal: [localhost]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python"}, "changed": false, "msg": "Key 'Name' was not found in the CSV header fields: ﻿Name, Configured-vCPU, ReclaimablevCPUs, ConfiguredMemory, ReclaimableMemory, ParentvCenter"}
```

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Open the file with encoding `utf-8-sig` to accommodate UTF8 BOM markers.

Fixes #544 

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
read_csv

### CODE REVIEW

File: changelogs/fragments/2592-read-csv-utf-bom.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - read_csv - opening file with encoding ``utf-8-sig`` (https://github.com/ansible-collections/community.general/issues/544).
```
Comment (felixfontein) [inline changelogs/fragments/2592-read-csv-utf-bom.yml:2]: The module_util is also used by at least one other plugin.

**Review by felixfontein — COMMENTED** [review]
I'm not sure whether this is really a bugfix, or more a feature. Maybe it's better to add a ``encoding`` option instead, so you can also specify `Latin1`? That's a pretty common encoding for CSV files created by Excel (at least in Europe).

**Comment by russoz** [comment]
Yup, the plot is thicker in this one. Will take a more thorough look later.

**Comment by Ajpantuso** [comment]
> Yup, the plot is thicker in this one. Will take a more thorough look later.

`to_native` is only there for the filter `from_csv` which is passed text from ansible.
In the case of `read_csv` the file is read as binary and unaffected by `to_native` altogether.
If you want to limit the scope to `read_csv` then just strip the BOM directly from the binary.

**Comment by felixfontein** [comment]
@russoz do you still plan to work on this one?

**Comment by russoz** [comment]
Yes but maybe taking a different approach. I think I'll kill this PR and raise a different one for that

**Pull request closed without merging**